src/lib/orcid.ts

```diff
@@ -0,0 +1,231 @@
+/**
+ * Module that handles ORCID authentication via OAuth2.
+ *
+ * "Three-legged OAuth" is used here. We originally tried using third-party libaries -- primarily vuex-oidc, which in
+ * turn uses oidc-client or oidc-client-ts for the actual OAuth2/OIDC authentication process. We ran into issues with
+ * all OAuth2 and OIDC authentication flows:
+ *
+ * - With OIDC's implicit flow, the client does not need to manage a client secret; this is good for single-page
+ *   applications whose source code is visible. This flow is supported by vuex-oidc v3 via oidc-client, though it is no
+ *   no longer supported by the supported branches of these projects. However, the access token has a limited lifespan,
+ *   and ORCID does not seem to support token renewal. (It does not return a renewal token, and its renew endpoint has
+ *   an X-Frame-Options header that prohibits use in an iframe, so in any case we cannot renew silently.) So the user is
+ *   logged out after 10 minutes.
+ * - The preferred option would be OIDC's authorization code flow with PKCE, which is more secure than the implicit flow
+ *   and still does not require a client secret. However, ORCID does not support PKCE. Plans to support it seem to have
+ *   stalled. (See https://github.com/ORCID/ORCID-Source/issues/5977.)
+ * - So we have opted to use three-legged authentication, which requires participation of the MaveDB API. In this flow,
+ *   the browser client first obtains a code when the user signs into ORCID and grants permissions. It sends this code
+ *   to the MaveDB API, which makes a second request to ORCID, this time exchanging the code and MaveDB's client secret
+ *   for an access token and, in addition, the ID token (JWT) that we will subsequently use for authorization. These
+ *   have a very long lifespan (which ORCID's OAuth2 service returns in the expires_in parameter, measured in seconds).
+ *   Three-leggged authentication is not supported by oidc-client or oidc-client-ts, so the present module implements
+ *   it.
+ *
+ * We could combine this module with the auth module that keeps track of user roles, but keeping the two distinct allows
+ * us to provide authentication methods other than ORCID.
+ *
+ * This module is not normally accessed directly by Vue components; instead, they should use lib/auth.ts.
+ */
+
+import axios from 'axios'
+import {v4 as uuidv4} from 'uuid'
+import {computed, ref, Ref} from 'vue'
+
+import config from '../config'
+
+export interface OidcUserProfileBase {
+  auth_time: number
+  family_name: string
+  given_name: string
+  id: string
+  jti: string
+  sub: string
+}
+
+// Notice the final '/'.
```

Comment:
not sure I understand the comment here - is this used other than with oidc-callback appended on?

Reply:
I went ahead and removed this unused code.